src/DSP/Reverter.cpp

```diff
@@ -0,0 +1,281 @@
+}
+
+inline void Reverter::switchBuffers() {
+    reverse_index = 0; 
+    // Reset the reverse index to start fresh for the new reverse playback segment.
+    
+    pos_start = pos_writer;
+    //
+
+    float pos_next = fmodf(float(pos_start + mem_size) - (reverse_index + phase_offset), mem_size); 
+    // Determine the position of the next sample (`pos_next`) after switching buffers:
+    // - Accounts for the current `reverse_index` (reset to 0 above) and phase offset.
+    // - Uses modulo (`fmodf`) to ensure the position wraps correctly within the circular buffer.
+    // The formula is the same as for pos_reader in updateReaderPosition()
+    
+    delta_crossfade = pos_reader - 1.0f - pos_next; 
+    // Calculate the sample distance between the current read position (`pos_reader`)
+    // and the newly computed `pos_next`:
+    // - Subtracting `1.0f` compensates for the fact that `pos_reader` corresponds to a read
+    //   position from the previous tick.
+
+    fade_counter = 0; 
+    // Reset the fade counter to begin a new crossfade for the next segment.
+}
+
+void Reverter::filterout(float *smp) {
+    writeToRingBuffer(smp);
+    processBuffer(smp);
+}
+
+void Reverter::writeToRingBuffer(const float *smp) {
+    int space_to_end = mem_size - pos_writer;
+    float abs_value = 0.0f;
+
+    if (buffersize <= space_to_end) {
+        // No wrap around, copy in one go
+        memcpy(&input[pos_writer], smp, buffersize * sizeof(float));
+    } else {
+        // Wrap around, split into two copies
+        memcpy(&input[pos_writer], smp, space_to_end * sizeof(float));
+        memcpy(&input[0], smp + space_to_end, (buffersize - space_to_end) * sizeof(float));
+    }
+    for (int i = 0; i < buffersize; i++) {
+        abs_value += fabsf(smp[i]);
+    }
+    // Update pos_writer after copying new buffer
+    pos_writer = (pos_writer + buffersize) % mem_size;
+    // calculate mean abs value of new buffer (for silence detection)
+    mean_abs_value = abs_value / static_cast<float>(buffersize);
+}
+
+void Reverter::processBuffer(float *smp) {
+    for (int i = 0; i < buffersize; i++) {
+        reverse_index++;
+        handleSync();
+        updateReaderPosition(i);
+        crossfadeSamples(smp, i);
+        applyGain(smp[i]);
+    }
+
+    phase_offset_old = phase_offset;
+    phase_offset_fade = 0.0f;
+}
+
+void Reverter::handleSync() {
+    switch (syncMode) {
+        case AUTO:
+            if (reverse_index >= delay) switchBuffers();
+            break;
+        case HOST:
+        case MIDI:
+            if (doSync && reverse_index >= syncPos) switchBuffers();
+            break;
+        case NOTEON:
+        case NOTEONOFF:
+            handleNoteSync();
+            
+            break;
+    }
+    // if everything fails
+    if (reverse_index >= max_delay && state == PLAYING) switchBuffers();
+}
+
+void Reverter::handleNoteSync() {
+    if (syncMode == NOTEON && reverse_index >= delay && state != IDLE) {
```

Comment:
recorded_samples is only needed in NOTEONOFF because it stores the recorded "duration" until the pseudo rms thresh is met. in the other modes the duration is known a priori because it is delay.